realm/realm-library/src/objectServer/java/io/realm/SyncSession.java

```diff
@@ -115,6 +136,105 @@ void notifySessionError(int errorCode, String errorMessage) {
         }
     }
 
+    // Called from native code
+    @SuppressWarnings("unused")
+    @KeepMember
+    synchronized void notifyProgressListener(long listenerId, long transferredBytes, long transferableBytes) {
+        Pair<ProgressListener, Progress> listener = listenerIdToProgressListenerMap.get(listenerId);
+        if (listener != null) {
+            Progress newProgressNotification = new Progress(transferredBytes, transferableBytes);
+            if (!newProgressNotification.equals(listener.second)) {
+                listener.first.onChange(newProgressNotification);
+                listener.second = newProgressNotification;
+            }
+        } else {
+            RealmLog.debug("Trying unknown listener failed: " + listenerId);
+        }
+    }
+    
+    /**
+     * Adds a progress listener tracking changes that need to be downloaded from the Realm Object
+     * Server.
+     * <p>
+     * The {@link ProgressListener} will be triggered immediately when registered, and periodically
+     * afterwards.
+     *
+     * @param mode type of mode used. See {@link ProgressMode} for more information.
+     * @param listener the listener to register.
+     */
+    public synchronized void addDownloadProgressListener(ProgressMode mode, ProgressListener listener) {
```

Comment:
I guess it is a tradeoff: Keep creating different monitor types which will be more fine-grained and complicated vs. just using `synchronized` which is more coarse-grained, but less complicated.

I don't see the SyncSession being used in a highly threaded manner, but if you feel strongly about it, I can introduce a specific monitor type?